cider-dap/src/adapter.rs

```diff
@@ -30,39 +31,88 @@ impl MyAdapter {
+    pub fn handle_breakpoint(
         &mut self,
         path: Source,
-        source: &Vec<SourceBreakpoint>,
+        points: &Vec<SourceBreakpoint>,
     ) -> Vec<Breakpoint> {
-        //Keep all the new breakpoints made
-        let mut out_vec: Vec<Breakpoint> = vec![];
+        // helper method to get diffs since it yelled at me about borrows when i had it in main method
+        fn calc_diffs(
+            new_set: &HashSet<i64>,
+            old_set: &HashSet<i64>,
+        ) -> (HashSet<i64>, HashSet<i64>) {
+            let to_set: HashSet<i64> =
+                new_set.difference(old_set).copied().collect();
+            let to_delete: HashSet<i64> =
+                old_set.difference(new_set).copied().collect();
+            (to_set, to_delete)
+        }
+
+        //check diffs
+        let mut new_point_set = HashSet::new();
+        for p in points {
+            new_point_set.insert(p.line);
+        }
+        let (to_set, to_delete) = calc_diffs(&new_point_set, &self.breakpoints);
+
+        //update adapter
+        self.breakpoints.clear();
+
+        let mut to_debugger_set: Vec<ParsedGroupName> = vec![];
+        let mut to_client: Vec<Breakpoint> = vec![];
+
+        // iterate over points recieved in request
+        for source_point in points {
+            self.breakpoints.insert(source_point.line);
+            let name = self.ids.lookup_line(source_point.line as u64);
 
-        //Loop over all breakpoints - why do we need to loop over all of them? is it bc input vec isnt mutable?
-        for source_point in source {
-            self.breakpoints.push((path.clone(), source_point.line));
-            //Create new Breakpoint instance
             let breakpoint = make_breakpoint(
                 self.break_count.increment().into(),
-                true,
+                name.is_some(),
                 Some(path.clone()),
                 Some(source_point.line),
             );
+            to_client.push(breakpoint);
 
-            out_vec.push(breakpoint);
+            if let Some((component, group)) = name {
+                if to_set.contains(&source_point.line) {
+                    to_debugger_set.push(ParsedGroupName::from_comp_and_group(
+                        component.clone(),
+                        group.clone(),
+                    ))
+                }
+            }
         }
-        //push breakpoints to cider debugger once have go ahead
-        out_vec
+        //send ones to set to debugger
+        self.debugger.set_breakpoints(to_debugger_set);
+        //delete from debugger
+        self.delete_breakpoints(to_delete);
+
+        //return list of created points to client
+        to_client
+    }
+    /// handles deleting breakpoints in the debugger
+    fn delete_breakpoints(&mut self, to_delete: HashSet<i64>) {
+        let mut to_debugger: Vec<ParsedGroupName> = vec![];
+        for point in to_delete {
```

Comment:
Since you're just iterating over the `HashSet`, you could just pass in an
iterator and avoid the `HashSet` altogether.